In [1]:
import pandas as pd
import numpy as np

In [2]:
with open('input.txt') as f:
    input_string = f.read().strip()

# Part 1

In [3]:
df = pd.DataFrame(list(map(int, list(input_string))), columns=['Input'])

In [4]:
df['Rolled'] = np.roll(df.Input, 1)

print (f'Part 1 answer: {df[df.Input == df.Rolled].Input.sum()}')

Part 1 answer: 1119


# Part 2

In [5]:
%%timeit
df['Rolled'] = np.roll(df.Input, int(len(df.Input)/2))

190 µs ± 5.85 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [8]:
print (f'Part 2 answer: {df[df.Input == df.Rolled].Input.sum()}')

Part 2 answer: 1420


### intel36
    223 µs ± 23.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
### dspython362
    190 µs ± 5.85 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)